File: frontend/src/types/proctoring.ts
// Proctoring Types for Enhanced AI Detection System

export type ProctoringEventType =
  // Browser-based events
  | 'tab_switch'
  | 'copy'
  | 'paste'
  | 'context_menu'
  | 'window_blur'
  // AI vision-based events
  | 'no_face_detected'
  | 'multiple_people'
  | 'phone_detected'
  | 'paper_detected'
  | 'looking_away';

export type ProctoringEventSeverity = 'low' | 'medium' | 'high' | 'critical';

export interface ProctoringEvent {
  classroom_quiz_id: number;
  event_type: ProctoringEventType;
  event_data?: string;
  attempt_id?: number;
  severity?: ProctoringEventSeverity;
  confidence?: number; // 0-100 for AI detections
}

export interface ProctoringEventResponse {
  event_id: number;
  warning_count: number;
  max_warnings: number;
  exceeded: boolean;
  severity: ProctoringEventSeverity;
  confidence?: number;
}

export interface DetectionStatus {
  faceDetected: boolean;
  faceCount: number;
  faceConfidence: number;
  multiplePeople: boolean;
  phoneDetected: boolean;
  paperDetected: boolean;
  bookDetected: boolean;
  lookingAway: boolean;
  detectedObjects: string[];
  violations: string[];
}

export interface ProctoringStats {
  total_events: number;
  events_by_type: Record<ProctoringEventType, number>;
  events_by_severity: Record<ProctoringEventSeverity, number>;
  critical_count: number;
  high_count: number;
}

File: backend/app/models/proctoring.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, Boolean
from sqlalchemy.orm import relationship
from app.core.time_utils import utcnow
from app.models.base import Base

class ProctoringEvent(Base):
    """Records suspicious activity during proctored quizzes.

    Tracks various proctoring violations including:
    - Browser behavior: tab switches, copy/paste, context menu, window blur
    - Face detection: no face, multiple people
    - Object detection: phones, papers/books
    - Attention: looking away from screen
    """
    __tablename__ = "proctoring_events"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), index=True, nullable=False)
    classroom_quiz_id = Column(Integer, ForeignKey("classroom_quizzes.id"), index=True, nullable=False)
    attempt_id = Column(Integer, ForeignKey("question_attempts.id", ondelete="SET NULL"), index=True, nullable=True)

    # Session token to group events from the same quiz attempt
    # This ensures violations are tracked per quiz attempt, not accumulated across all attempts
    session_token = Column(String, index=True, nullable=True)

    # Event type - expanded to include AI proctoring events
    # Browser: "tab_switch", "copy", "paste", "context_menu", "window_blur"
    # AI Vision: "no_face_detected", "multiple_people", "phone_detected",
    #            "paper_detected", "looking_away"
    event_type = Column(String, index=True, nullable=False)

    # Additional context about the event
    event_data = Column(Text, nullable=True)

    # Severity level: "low", "medium", "high", "critical"
    # Helps teachers prioritize which violations to review
    severity = Column(String, default="medium", nullable=False)

    # AI confidence score (0.0 - 1.0) for vision-based detections
    confidence = Column(Integer, nullable=True)  # Stored as int (0-100)

    # Whether this was a false positive (teacher can mark after review)
    is_false_positive = Column(Boolean, default=False, nullable=False)

    # Timestamp when the event occurred
    timestamp = Column(DateTime, default=utcnow, index=True, nullable=False)

    # Relationships
    user = relationship("User")
    quiz = relationship("ClassroomQuiz", back_populates="proctoring_events")
